watcher/watcher.go

```diff
@@ -2,106 +2,157 @@ package watcher
+	refCh         chan string
+}
 
-	//	(Optional) Context to use for stopping of watcher.
-	context context.Context
+func NewGitWatcher(status *util.Status, log logrus.FieldLogger) *GitWatcher {
+	return &GitWatcher{
+		status:        status,
+		log:           log,
+		repoCreatedCh: make(chan bool),
+		branchCh:      make(chan string),
+		refCh:         make(chan string),
+		repoExists:    util.PathExists(nhost.GIT_DIR),
+	}
 }
 
-//	Add individial location to watcher.
-//	Along with associating it with respectiove operation function.
-func (w *Watcher) Register(path string, op Operation) error {
+func (gw *GitWatcher) Watch(ctx context.Context, interval time.Duration, reloadFunc func(branch, ref string) error) {
+	go gw.watchRepoExists(ctx, interval)
+	go gw.watchBranchChange(ctx, interval)
+	go gw.watchRefChange(ctx, interval)
 
-	w.log.WithField("component", "path").Debugln("Watching", util.Rel(path))
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case _, ok := <-gw.repoCreatedCh:
+			gw.status.Infoln("Git repo created")
+			if !ok {
+				// nullify the channel, so it's not ready for communication and we don't fall into an infinite loop
+				gw.repoCreatedCh = nil
+			}
+		case branch := <-gw.branchCh:
+			gw.status.Infoln(fmt.Sprintf("Detected branch change: %s", branch))
+			gw.log.WithField("branch", branch).Debug("Detected branch change")
 
-	w.Map[path] = op
+			if err := reloadFunc(gw.branch, gw.remoteRef); err != nil {
+				gw.log.WithError(err).Errorln("Failed to reload")
+				gw.status.Errorln(fmt.Sprintf("Failed to reload: %v", err))
+			}
+		case ref := <-gw.refCh:
+			gw.status.Infoln(fmt.Sprintf("Detected remoteRef change: %s", ref))
+			gw.log.WithField("ref", ref[:7]).Debug("Detected git remote ref change")
 
-	return w.Add(path)
+			if err := reloadFunc(gw.branch, gw.remoteRef); err != nil {
+				gw.log.WithError(err).Errorln("Failed to reload")
+				gw.status.Errorln(fmt.Sprintf("Failed to reload: %v", err))
+			}
+		}
+	}
 }
 
-//	Validates whether a given key is already
-//	register in the watcher.
-func (w *Watcher) Registered(key string) bool {
+func (gw *GitWatcher) watchBranchChange(ctx context.Context, interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	for range ticker.C {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if !util.PathExists(filepath.Join(nhost.GIT_DIR, "HEAD")) {
+				continue
+			}
+
+			branch := nhost.GetCurrentBranch()
+			if gw.branch == "" {
+				// if the branch is not set yet, do not notify to prevent reload triggering
+				gw.branch = branch
+				continue
+			}
 
-	for name := range w.Map {
-		if name == key {
-			return true
+			if gw.branch != branch {
+				gw.branch = branch
+				gw.branchCh <- branch
+			}
 		}
 	}
-
-	return false
 }
 
-//	Initializes a new watcher
-func New(ctx context.Context) *Watcher {
+func (gw *GitWatcher) watchRefChange(ctx context.Context, interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	for range ticker.C {
```

Comment:
I'd suggest putting the ticket inside the select:

```
for {
  select {
 	 	case <-ctx.Done():
 		 	return  
   case <- ticker.C:
      // here the code in default
}
```

Main reason is that otherwise you are blocking the entire thing for `interval`. Even if the context is cancelled you still have to wait as you don't evaluate the select until then.